typescript/README.md

```diff
@@ -10,31 +53,279 @@ npm install @dfinity/llm
+
+**Common use cases for tools:**
+- **Data retrieval**: Fetching real-time information (prices, weather, account balances)
+- **External API calls**: Integrating with third-party services
+- **Calculations**: Performing complex computations
+- **Database operations**: Querying or updating data
+- **Custom business logic**: Executing domain-specific functions
+
+**How it works:**
+1. You define available tools with their parameters
+2. The LLM analyzes the user's request
+3. If a tool would be helpful, the LLM returns a "tool call" instead of a direct answer
+4. Your code executes the requested tool with the LLM's provided parameters
+5. You send the tool's result back to the LLM
+6. The LLM incorporates the result into its final response
+
+#### Defining and Using a Tool
+
+You can define tools that the LLM can use to perform actions:
+
+```typescript
+import { IDL, update } from "azle";
+import * as llm from "@dfinity/llm";
+
+export default class {
+  @update([], llm.Response)
+  async example(): Promise<llm.Response> {
+    return await llm.chat(llm.Model.Llama3_1_8B)
+      .withMessages([
+        llm.createSystemMessage("You are a helpful assistant"),
+        llm.createUserMessage("What's the balance of account abc123?")
+      ])
+      .withTools([
+        llm.tool("icp_account_balance")
+          .withDescription("Lookup the balance of an ICP account")
+          .withParameter(
+            llm.parameter("account", llm.ParameterType.String)
+              .withDescription("The ICP account to look up")
+              .isRequired()
+          )
+          .build()
+      ])
+      .send();
   }
 }
 ```
 
-### Chatting (multiple messages)
+#### Handling Tool Calls from the LLM
+
+When the LLM decides to use one of your tools, you can handle the call:
 
 ```typescript
 import { IDL, update } from "azle";
-import { chat_message as ChatMessageIDL } from "azle/canisters/llm/idl";
 import * as llm from "@dfinity/llm";
 
 export default class {
-  @update([IDL.Vec(ChatMessageIDL)], IDL.Text)
-  async chat(messages: llm.ChatMessage[]): Promise<string> {
-    return await llm.chat(llm.Model.Llama3_1_8B, messages);
+  @update([], llm.Response)
+  async example(): Promise<llm.Response> {
+    const response = await llm.chat(llm.Model.Llama3_1_8B)
+      .withMessages([
+        llm.createSystemMessage("You are a helpful assistant"),
+        llm.createUserMessage("What's the weather in San Francisco?")
+      ])
+      .withTools([
+        llm.tool("get_weather")
+          .withDescription("Get current weather for a location")
+          .withParameter(
+            llm.parameter("location", llm.ParameterType.String)
+              .withDescription("The location to get weather for")
+              .isRequired()
+          )
+          .build()
```

Comment:
Same here. Build patterns are generally not as necessary in TS, because you can define interfaces with optional fields and with default values:

```
    {
      name: "get_weather",
      description: "Get current weather for a location",
      parameters: [
        {
          name: "location",
          type: "string",
          description: "The location to get weather for",
          required: true
        }
      ]
    }
```